discord/automod.py

```diff
@@ -0,0 +1,420 @@
+            raise ValueError('presets must be set if type is keyword_preset.')
+
+    def __repr__(self) -> str:
+        return f"<AutoModTrigger type={self.type}>"
+
+    @classmethod
+    def from_data(cls: Type[Self], type: int, data: Optional[AutoModerationTriggerMetadataPayload]) -> Self:
+        type_ = try_enum(AutoModRuleTriggerType, type)
+        if type_ is AutoModRuleTriggerType.keyword:
+            return cls(type=type_, keyword_filter=data['keyword_filter'])  # type: ignore # unable to typeguard due to outer payload
+        elif type_ is AutoModRuleTriggerType.keyword_preset:
+            return cls(type=type_, presets=AutoModPresets._from_value(data['presets']))  # type: ignore # unable to typeguard due to outer payload
+        return cls(type=type_)
+
+    def to_metadata_dict(self) -> AutoModerationTriggerMetadataPayload:
+        if self.type is AutoModRuleTriggerType.keyword:
+            return {'keyword_filter': self.keyword_filter}  # type: ignore # guarded by the type check
+        elif self.type is AutoModRuleTriggerType.keyword_preset:
+            return {'presets': self.presets.value}  # type: ignore # guarded by the type check
+
+        return {}
+
+
+class AutoModRule:
+    """Represents an auto moderation rule.
+
+    Attributes
+    -----------
+    id: :class:`int`
+        The ID of the rule.
+    guild: :class:`Guild`
+        The guild the rule is for.
+    name: :class:`str`
+        The name of the rule.
+    creator_id: :class:`int`
+        The ID of the user that created the rule.
+    trigger: :class:`AutoModTrigger`
+        The rule's trigger.
+    enabled: :class:`bool`
+        Whether the rule is enabled.
+    raw_exempt_roles: List[:class:`int`]
+        The IDs of the roles that are exempt from the rule.
+    raw_exempt_channels: List[:class:`int`]
+        The IDs of the channels that are exempt from the rule.
+
+    .. versionadded:: 2.0
+    """
+
+    __slots__ = (
+        '_state',
+        'id',
+        'guild',
+        'name',
+        'creator_id',
+        'event_type',
+        'trigger',
+        'enabled',
+        'raw_exempt_roles',
+        'raw_exempt_channels',
+        '_actions',
+    )
+
+    def __init__(self, *, data: AutoModerationRulePayload, guild: Guild, state: ConnectionState) -> None:
+        self._state: ConnectionState = state
+        self.guild: Guild = guild
+        self.id: int = int(data['id'])
+        self.name: str = data['name']
+        self.creator_id = int(data['creator_id'])
+        self.event_type: AutoModRuleEventType = try_enum(AutoModRuleEventType, data['event_type'])
+        self.trigger: AutoModTrigger = AutoModTrigger.from_data(data['trigger_type'], data=data.get('trigger_metadata'))
+        self.enabled: bool = data['enabled']
+        self.raw_exempt_roles: List[int] = [int(role_id) for role_id in data['exempt_roles']]
+        self.raw_exempt_channels: List[int] = [int(channel_id) for channel_id in data['exempt_channels']]
```

Comment:
```suggestion
        self.exempt_role_ids: List[int] = [int(role_id) for role_id in data['exempt_roles']]
        self.exempt_channel_ids: List[int] = [int(channel_id) for channel_id in data['exempt_channels']]
```

I wonder if casting these into a set makes more sense than list here since it allows for quicker O(1) checking. It also allows there to be a quick `AutoModRule.is_exempt` utility method.